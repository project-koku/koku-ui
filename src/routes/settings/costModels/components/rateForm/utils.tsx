import { SortByDirection } from '@patternfly/react-table';
import type { CostModel, CostModelRequest } from 'api/costModels';
import type { MetricHash } from 'api/metrics';
import type { Rate, RateRequest, TagRates } from 'api/rates';
import { countDecimals, formatCurrencyRateRaw, isCurrencyFormatValid, unFormat } from 'utils/format';

import { textHelpers } from './constants';

export const initialTaggingRates = {
  tagKey: {
    isDirty: false,
    value: '',
  },
  defaultTag: null,
  tagValues: [
    {
      description: '',
      isDirty: false,
      isTagValueDirty: false,
      tagValue: '',
      value: '',
    },
  ],
};

export const initialRateFormData = {
  otherTiers: [] as Rate[],
  step: 'initial',
  description: '',
  metric: '',
  measurement: {
    value: '',
    isDirty: false,
  },
  calculation: '',
  rateKind: 'regular',
  tieredRates: [
    {
      isDirty: false,
      value: '',
    },
  ],
  taggingRates: { ...initialTaggingRates },
  errors: {
    description: null,
    measurement: textHelpers.required,
    tieredRates: textHelpers.required,
    tagValues: [textHelpers.required],
    tagDescription: [null],
    tagKey: textHelpers.required,
    tagValueValues: [textHelpers.required],
  },
};

export type RateFormData = typeof initialRateFormData;
export type RateFormTagValue = (typeof initialRateFormData)['taggingRates']['tagValues'][0];
export type taggingRates = (typeof initialRateFormData)['taggingRates'];
export type RateFormErrors = (typeof initialRateFormData)['errors'];

export const checkRateOnChange = (value: string) => {
  if (value.length === 0) {
    return textHelpers.required;
  }
  if (!isCurrencyFormatValid(value)) {
    return textHelpers.not_number;
  }
  if (Number(value) < 0) {
    return textHelpers.not_positive;
  }
  // Test number of decimals
  const decimals = countDecimals(value);
  if (decimals > 10) {
    return textHelpers.rate_too_long;
  }
  return null;
};

export function getDefaultCalculation(metricsHash: MetricHash, metric: string) {
  let options = Object.keys(metricsHash);
  if (!options.includes(metric)) {
    return '';
  }
  options = Object.keys(metricsHash[metric]);
  if (options.length === 0) {
    return '';
  }
  return metricsHash[metric][options[0]].default_cost_type;
}

export function genFormDataFromRate(rate: Rate, defaultValue = initialRateFormData, tiers: Rate[]): RateFormData {
  const otherTiers = tiers || defaultValue.otherTiers;

  if (!rate) {
    return { ...defaultValue, otherTiers };
  }
  const rateKind = rate.tiered_rates ? 'regular' : 'tagging';
  let tieredRates = [{ isDirty: true, value: '' }];
  const tagRates = { ...initialTaggingRates };
  const errors = {
    description: null,
    measurement: null,
    tieredRates: null,
    tagValues: [null],
    tagKey: null,
    tagValueValues: [null],
    tagDescription: [null],
  };
  if (rateKind === 'tagging') {
    const item = rate.tag_rates as TagRates;
    tagRates.tagKey = { isDirty: true, value: item.tag_key };
    const defaultIndex = item.tag_values.findIndex(tvalue => tvalue.default);
    tagRates.defaultTag = defaultIndex === -1 ? null : defaultIndex;
    tagRates.tagValues = item.tag_values.map(tvalue => {
      const value = formatCurrencyRateRaw(tvalue.value, tvalue.unit);
      return {
        description: tvalue.description,
        isDirty: false,
        isTagValueDirty: false,
        tagValue: tvalue.tag_value,
        value,
      };
    });
    errors.tieredRates = textHelpers.required;
    errors.tagValueValues = new Array(item.tag_values.length).fill(null);
    errors.tagValues = new Array(item.tag_values.length).fill(null);
    errors.tagDescription = new Array(item.tag_values.length).fill(null);
  }
  if (rateKind === 'regular') {
    tieredRates = rate.tiered_rates.map(tieredRate => {
      const value = formatCurrencyRateRaw(tieredRate.value, tieredRate.unit);
      return {
        isDirty: true,
        value,
      };
    });
    errors.tagValues = [textHelpers.required];
    errors.tagValueValues = [textHelpers.required];
  }
  return {
    otherTiers,
    step: 'set_rate',
    description: rate.description,
    metric: rate.metric.label_metric,
    measurement: {
      value: rate.metric.name,
      isDirty: true,
    },
    calculation: rate.cost_type,
    rateKind,
    tieredRates,
    taggingRates: tagRates,
    errors,
  };
}

export const mergeToRequest = (
  metricsHash: MetricHash,
  costModel: CostModel,
  rateFormData: RateFormData,
  index: number = -1
): CostModelRequest => {
  if (index < 0) {
    index = costModel.rates.length;
  }
  const rate = transformFormDataToRequest(rateFormData, metricsHash, costModel.currency) as RateRequest;
  return {
    currency: costModel.currency,
    name: costModel.name,
    source_type: 'OCP',
    description: costModel.description,
    distribution_info: {
      distribution_type: costModel.distribution_info ? costModel.distribution_info.distribution_type : undefined,
      network_unattributed: costModel.distribution_info ? costModel.distribution_info.network_unattributed : undefined,
      platform_cost: costModel.distribution_info ? costModel.distribution_info.platform_cost : undefined,
      storage_unattributed: costModel.distribution_info ? costModel.distribution_info.storage_unattributed : undefined,
      worker_cost: costModel.distribution_info ? costModel.distribution_info.worker_cost : undefined,
    },
    source_uuids: costModel.sources.map(src => src.uuid),
    markup: { value: costModel.markup.value, unit: 'percent' },
    rates: [...costModel.rates.slice(0, index), rate, ...costModel.rates.slice(index + 1)],
  };
};

export const transformFormDataToRequest = (
  rateFormData: RateFormData,
  metricsHash: MetricHash,
  currencyUnits: string = 'USD'
): Rate => {
  const ratesKey = rateFormData.rateKind === 'tagging' ? 'tag_rates' : 'tiered_rates';
  const ratesBody =
    rateFormData.rateKind === 'tagging'
      ? {
          tag_key: rateFormData.taggingRates.tagKey.value,
          tag_values: rateFormData.taggingRates.tagValues.map((tvalue, ix) => {
            return {
              tag_value: tvalue.tagValue,
              unit: currencyUnits,
              value: unFormat(tvalue.value), // Normalize for API requests where USD decimal format is expected
              description: tvalue.description,
              default: ix === rateFormData.taggingRates.defaultTag,
            };
          }),
        }
      : rateFormData.tieredRates.map(tiered => {
          return {
            value: unFormat(tiered.value), // Normalize for API requests where USD decimal format is expected
            unit: currencyUnits,
            usage: { unit: currencyUnits },
          };
        });
  const metricData = metricsHash[rateFormData.metric][rateFormData.measurement.value];
  return {
    description: rateFormData.description,
    metric: {
      metric: metricData.metric,
      name: metricData.metric,
      label_metric: metricData.label_metric,
      label_measurement: metricData.label_measurement,
      label_measurement_unit: metricData.label_measurement_unit,
      source_type: 'OpenShift Cluster Platform',
      default_cost_type: metricData.default_cost_type,
    },
    cost_type: rateFormData.calculation,
    [ratesKey]: ratesBody,
  };
};

export interface OtherTier {
  metric: RateFormData['metric'];
  measurement: RateFormData['measurement']['value'];
  costType: RateFormData['calculation'];
  tagKey: RateFormData['taggingRates']['tagKey']['value'];
}

export const OtherTierFromRate = (rate: Rate): OtherTier => {
  const tagKey = rate.tag_rates && rate.tag_rates.tag_key ? rate.tag_rates.tag_key : null;
  return {
    metric: rate.metric.label_metric,
    measurement: rate.metric.name,
    tagKey,
    costType: rate.cost_type,
  };
};

export const OtherTierFromRateForm = (rateData: RateFormData): OtherTier => {
  const tagKey = rateData.taggingRates && rateData.taggingRates.tagKey ? rateData.taggingRates.tagKey.value : null;
  const res = {
    metric: rateData.metric,
    measurement: rateData.measurement ? rateData.measurement.value : null,
    tagKey,
    costType: rateData.calculation,
  };
  return res;
};

export const isDuplicateTagRate = (rate: OtherTier, current: OtherTier) => {
  return (
    rate.metric === current.metric &&
    rate.measurement === current.measurement &&
    rate.costType === current.costType &&
    rate.tagKey === current.tagKey
  );
};

export type CompareResult = 1 | -1 | 0;

export function compareBy(
  r1: Rate,
  r2: Rate,
  direction: keyof typeof SortByDirection,
  projection: (r: Rate) => string
): CompareResult {
  const m1 = projection(r1);
  const m2 = projection(r2);
  if (direction === SortByDirection.asc) {
    return m1 > m2 ? 1 : m1 < m2 ? -1 : 0;
  }
  return m1 > m2 ? -1 : m1 < m2 ? 1 : 0;
}

export const descriptionErrors = (value: string) => {
  if (value.length > 500) {
    return textHelpers.description_too_long;
  }
  return null;
};

export const tagKeyValueErrors = (value: string) => {
  if (value.length === 0) {
    return textHelpers.required;
  }
  if (value.length > 100) {
    return textHelpers.tag_too_long;
  }
  return null;
};
