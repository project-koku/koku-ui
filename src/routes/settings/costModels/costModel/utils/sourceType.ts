export const getSourceType = (sourceType: string) => {
  let result;
  switch (sourceType) {
    case 'Amazon Web Services':
      result = 'AWS';
      break;
    case 'Google Cloud':
      result = 'GCP';
      break;
    case 'Microsoft Azure':
      result = 'Azure';
      break;
    case 'OpenShift Container Platform':
      result = 'OCP';
      break;
  }
  return result;
};
