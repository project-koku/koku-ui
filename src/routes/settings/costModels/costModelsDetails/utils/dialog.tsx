import {
  Alert,
  AlertVariant,
  Button,
  ButtonVariant,
  Content,
  ContentVariants,
  Stack,
  StackItem,
} from '@patternfly/react-core';
import React from 'react';

export function DeleteDialogActions({
  status,
  deleteText,
  deleteAction,
  cancelText,
  cancelAction,
  sourcesNo,
}): JSX.Element[] {
  let actions = [];
  if (status === 'loading') {
    actions = [
      <Button key="pending_delete" variant={ButtonVariant.danger} isDisabled>
        {deleteText}
      </Button>,
      <Button key="pending_cancel" variant={ButtonVariant.link} isDisabled>
        {cancelText}
      </Button>,
    ];
  }
  if (['failure', 'open'].includes(status) && sourcesNo > 0) {
    actions = [
      <Button key="sources_cancel" variant={ButtonVariant.link} onClick={cancelAction}>
        {cancelText}
      </Button>,
    ];
  }
  if (['failure', 'open'].includes(status) && sourcesNo === 0) {
    actions = [
      <Button key="delete" variant={ButtonVariant.danger} onClick={deleteAction}>
        {deleteText}
      </Button>,
      <Button key="cancel" variant={ButtonVariant.link} onClick={cancelAction}>
        {cancelText}
      </Button>,
    ];
  }
  return actions;
}

interface CannotDeleteContentProps {
  head: string;
  body: string;
  sources: string[];
}

function CannotDeleteContent({ head, body, sources }: CannotDeleteContentProps): JSX.Element {
  return (
    <Stack hasGutter>
      <StackItem>
        <Content>
          <Content component="p">{head}</Content>
        </Content>
      </StackItem>
      <StackItem>
        <Content>
          <Content component="p">{body}</Content>
          <Content component={ContentVariants.ol}>
            {sources.map(source => (
              <Content component="li" key={source}>
                {source}
              </Content>
            ))}
          </Content>
        </Content>
      </StackItem>
    </Stack>
  );
}

export function DeleteDialogBody({
  status,
  sources,
  error,
  cannotDeleteTitle,
  cannotDeleteBody,
  canDeleteBody,
}): JSX.Element {
  if (status !== 'close' && sources.length > 0) {
    return <CannotDeleteContent head={cannotDeleteTitle} body={cannotDeleteBody} sources={sources} />;
  }
  if (status !== 'close' && sources.length === 0) {
    return (
      <Stack hasGutter>
        {status === 'failure' ? (
          <StackItem>
            <Alert variant={AlertVariant.danger} title={error} />
          </StackItem>
        ) : null}
        <StackItem>{canDeleteBody}</StackItem>
      </Stack>
    );
  }
}

export function getDialogStateName(isLoading: boolean, isOpen: boolean, error: string): string {
  if (!isOpen) {
    return 'close';
  }
  if (isLoading) {
    return 'loading';
  }
  if (error) {
    return 'failure';
  }
  return 'open';
}
