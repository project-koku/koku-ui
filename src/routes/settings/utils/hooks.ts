import { useAddNotification } from '@redhat-cloud-services/frontend-components-notifications/hooks';
import type { AccountSettingsType } from 'api/accountSettings';
import type { SettingsType } from 'api/settings';
import type { AxiosError } from 'axios';
import { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import type { AnyAction } from 'redux';
import type { ThunkDispatch } from 'redux-thunk';
import type { RootState } from 'store';
import { accountSettingsSelectors } from 'store/accountSettings';
import { FetchStatus } from 'store/common';
import { settingsSelectors } from 'store/settings';
import { resetStatus } from 'store/settings/settingsActions';

interface AccountSettingsUpdateProps<T> {
  type: AccountSettingsType;
  getSessionValue: () => T;
  setState: (v: T) => void;
}

interface SettingsUpdateProps {
  type: SettingsType;
}

interface SettingsNotificationProps {
  error: AxiosError;
  notification: Notification;
  status: FetchStatus;
}

export const useAccountSettingsUpdate = <T>({
  type,
  getSessionValue,
  setState,
}: AccountSettingsUpdateProps<T>): SettingsNotificationProps => {
  const dispatch = useDispatch<ThunkDispatch<RootState, any, AnyAction>>();
  const addNotification = useAddNotification();

  const error = useSelector((state: RootState) =>
    accountSettingsSelectors.selectAccountSettingsUpdateError(state, type)
  ) as AxiosError | undefined;
  const notification = useSelector((state: RootState) =>
    accountSettingsSelectors.selectAccountSettingsUpdateNotification(state, type)
  );
  const status = useSelector((state: RootState) =>
    accountSettingsSelectors.selectAccountSettingsUpdateStatus(state, type)
  );

  useEffect(() => {
    if (status === FetchStatus.complete) {
      if (!error) {
        setState(getSessionValue());
      }
      if (notification) {
        addNotification(notification as any);
        dispatch(resetStatus());
      }
    }
  }, [addNotification, dispatch, error, getSessionValue, notification, setState, status]);

  return { error, notification, status };
};

export const useSettingsUpdate = ({ type }: SettingsUpdateProps): SettingsNotificationProps => {
  const dispatch = useDispatch<ThunkDispatch<RootState, any, AnyAction>>();
  const addNotification = useAddNotification();

  const error = useSelector((state: RootState) => settingsSelectors.selectSettingsUpdateError(state, type)) as
    | AxiosError
    | undefined;
  const notification = useSelector((state: RootState) =>
    settingsSelectors.selectSettingsUpdateNotification(state, type)
  );
  const status = useSelector((state: RootState) => settingsSelectors.selectSettingsUpdateStatus(state, type));

  useEffect(() => {
    if (status === FetchStatus.complete && notification) {
      addNotification(notification as any);
      dispatch(resetStatus());
    }
  }, [addNotification, dispatch, error, notification, status]);

  return { error, notification, status };
};
