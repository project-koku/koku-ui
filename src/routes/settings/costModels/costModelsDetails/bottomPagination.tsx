import type { PaginationProps } from '@patternfly/react-core';
import { Pagination, PaginationVariant } from '@patternfly/react-core';
import { intl as defaultIntl } from 'components/i18n';
import messages from 'locales/messages';
import React from 'react';
import type { WrappedComponentProps } from 'react-intl';
import { injectIntl } from 'react-intl';
import { connect } from 'react-redux';
import type { RootState } from 'store';
import { costModelsSelectors } from 'store/costModels';
import type { RouterComponentProps } from 'utils/router';
import { withRouter } from 'utils/router';

import { initialCostModelsQuery, limitTransform, offsetTransform, stringifySearch } from './utils/query';

type BottomPaginationBaseProps = Omit<PaginationProps, 'ref'> & WrappedComponentProps;

const BottomPaginationBase: React.FC<BottomPaginationBaseProps> = props => {
  const {
    variant,
    intl = defaultIntl, // for testing
    itemCount,
    page,
    perPage,
    onSetPage,
    onPerPageSelect,
  } = props;
  return (
    <Pagination
      onPerPageSelect={onPerPageSelect}
      onSetPage={onSetPage}
      variant={variant}
      itemCount={itemCount}
      page={page}
      perPage={perPage}
      titles={{
        paginationAriaLabel: intl.formatMessage(messages.paginationTitle, {
          title: intl.formatMessage(messages.costModelsDetailsTitle),
          placement: 'bottom',
        }),
      }}
    />
  );
};

const mapStateToProps = (state: RootState) => {
  const { count, page, perPage } = costModelsSelectors.pagination(state);
  const query: { [p: string]: string } = costModelsSelectors.query(state);
  return {
    query,
    count,
    page,
    perPage,
  };
};

const mergeProps = (stateProps: ReturnType<typeof mapStateToProps>, dispatchProps, ownProps: RouterComponentProps) => {
  const { router } = ownProps;
  const { count, page, perPage, query } = stateProps;
  return {
    itemCount: count,
    page,
    perPage,
    variant: PaginationVariant.bottom,
    onSetPage: (_evt: React.MouseEvent | React.KeyboardEvent | MouseEvent, newPage: number, newPerPage: number) => {
      router.navigate(
        stringifySearch({
          ...initialCostModelsQuery,
          ...query,
          offset: offsetTransform(newPage, newPerPage),
        })
      );
    },
    onPerPageSelect: (_evt: React.MouseEvent | React.KeyboardEvent | MouseEvent, newPerPage: number) => {
      router.navigate(
        stringifySearch({
          ...initialCostModelsQuery,
          ...query,
          limit: limitTransform(newPerPage),
          offset: 0,
        })
      );
    },
  };
};

export const CostModelsBottomPagination = withRouter(
  connect(mapStateToProps, undefined, mergeProps)(injectIntl(BottomPaginationBase))
);
