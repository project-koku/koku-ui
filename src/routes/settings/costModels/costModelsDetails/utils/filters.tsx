import type { ToolbarProps } from '@patternfly/react-core';
import { SearchInput } from '@patternfly/react-core';
import { Toolbar, ToolbarFilter } from '@patternfly/react-core';
import { intl as defaultIntl } from 'components/i18n';
import messages from 'locales/messages';
import React, { useState } from 'react';
import type { WrappedComponentProps } from 'react-intl';
import { injectIntl } from 'react-intl';
import { connect } from 'react-redux';
import type { SelectWrapperOption } from 'routes/components/selectWrapper';
import { SelectWrapper } from 'routes/components/selectWrapper';
import type { RootState } from 'store';
import { costModelsSelectors } from 'store/costModels';
import type { RouterComponentProps } from 'utils/router';
import { withRouter } from 'utils/router';

import type { CostModelsQuery } from './query';
import { initialCostModelsQuery, stringifySearch } from './query';
import type { Inputer } from './types';

interface FilterInputProps {
  value: string;
  onChange: (value: string) => void;
  onSearch: (event, value: string) => void;
  placeholder?: string;
}

const FilterInput: React.FC<FilterInputProps> = ({ placeholder = '', value, onChange, onSearch }) => {
  return (
    <SearchInput
      aria-label={placeholder}
      id="cost-model-filter-input"
      onChange={(_evt, val) => onChange(val)}
      onClear={() => onChange('')}
      onSearch={onSearch}
      placeholder={placeholder}
      value={value}
    />
  );
};

export const onSearch = (router, key: string, query: CostModelsQuery, inputer: Inputer) => () => {
  if (inputer.value !== '') {
    const currentValue = inputer.value;
    router.navigate(
      stringifySearch({
        ...query,
        [key]: query[key] ? `${query[key]},${currentValue}` : currentValue,
      })
    );
    inputer.setValue('');
  }
};

export const onDeleteChip = (router, key, query: CostModelsQuery) => {
  return (_filterName: string, chipName: string) => {
    const newState = query[key] ? query[key].split(',').filter(qval => qval !== chipName) : null;
    router.navigate(
      stringifySearch({
        ...query,
        [key]: newState === null || newState.length === 0 ? null : (newState.join(',') as string),
      })
    );
  };
};

export const onDeleteChipGroup = (router, query: CostModelsQuery, key: string) => {
  return () => {
    router.navigate(stringifySearch({ ...query, [key]: null }));
  };
};

interface DescriptionFilterOwnProps {
  deleteChip?: any;
  deleteChipGroup?: any;
  filterType?: any;
  chips?: any;
  categoryName?: string;
  query?: any;
}

type DescriptionFilterProps = DescriptionFilterOwnProps & WrappedComponentProps & RouterComponentProps;

const descriptionMapStateToProps = (state: RootState) => {
  const filterType = costModelsSelectors.currentFilterType(state);
  const query: Partial<CostModelsQuery> = costModelsSelectors.query(state);
  return { filterType, query };
};

const descriptionMergeProps = (
  stateProps: ReturnType<typeof descriptionMapStateToProps>,
  dispatchProps,
  ownProps: WrappedComponentProps & RouterComponentProps
) => {
  const {
    intl = defaultIntl, // Default required for testing
    router,
  } = ownProps;
  const { filterType, query } = stateProps;

  const chips = query.description ? query.description.split(',') : [];
  return {
    categoryName: intl.formatMessage(messages.description),
    chips,
    deleteChip: onDeleteChip(router, 'description', { ...initialCostModelsQuery, ...query }),
    deleteChipGroup: onDeleteChipGroup(router, { ...initialCostModelsQuery, ...query }, 'description'),
    filterType,
    intl,
    query,
    router,
  };
};

const DescriptionFilterBase: React.FC<DescriptionFilterProps> = ({
  deleteChip,
  deleteChipGroup,
  filterType,
  intl,
  chips,
  categoryName,
  query,
  router,
}) => {
  const [value, setValue] = React.useState('');
  const children =
    filterType === 'description' ? (
      <FilterInput
        placeholder={intl.formatMessage(messages.filterByPlaceholder, { value: filterType }) as string}
        value={value}
        onChange={(val: string) => setValue(val)}
        onSearch={onSearch(router, 'description', { ...initialCostModelsQuery, ...query }, { value, setValue })}
      />
    ) : null;
  return (
    <ToolbarFilter
      deleteLabel={deleteChip}
      deleteLabelGroup={deleteChipGroup}
      labels={chips}
      categoryName={categoryName}
    >
      {children}
    </ToolbarFilter>
  );
};
const DescriptionFilterConnect = connect(
  descriptionMapStateToProps,
  undefined,
  descriptionMergeProps
)(DescriptionFilterBase);
export const DescriptionFilter = injectIntl(withRouter(DescriptionFilterConnect));

interface NameFilterOwnProps {
  deleteChip?: any;
  deleteChipGroup?: any;
  filterType?: any;
  chips?: any;
  categoryName?: string;
  query?: any;
}

type NameFilterProps = NameFilterOwnProps & WrappedComponentProps & RouterComponentProps;

const nameFilterMapStateToProps = (state: RootState) => {
  const filterType = costModelsSelectors.currentFilterType(state);
  const query: Partial<CostModelsQuery> = costModelsSelectors.query(state);
  return { filterType, query };
};

const nameFilterMergeProps = (
  stateProps: ReturnType<typeof nameFilterMapStateToProps>,
  dispatchProps,
  ownProps: WrappedComponentProps & RouterComponentProps
) => {
  const {
    intl = defaultIntl, // Default required for testing
    router,
  } = ownProps;

  const { filterType, query } = stateProps;
  const chips = query.name ? query.name.split(',') : [];

  return {
    deleteChip: onDeleteChip(router, 'name', { ...initialCostModelsQuery, ...query }),
    deleteChipGroup: onDeleteChipGroup(router, { ...initialCostModelsQuery, ...query }, 'name'),
    filterType,
    chips,
    categoryName: intl.formatMessage(messages.names, { count: 1 }),
    intl,
    query,
    router,
  };
};

const NameFilterBase: React.FC<NameFilterProps> = ({
  deleteChip,
  deleteChipGroup,
  filterType,
  intl,
  chips,
  categoryName,
  query,
  router,
}) => {
  const [value, setValue] = React.useState('');
  const children =
    filterType === 'name' ? (
      <FilterInput
        placeholder={intl.formatMessage(messages.filterByPlaceholder, { value: filterType }) as string}
        value={value}
        onChange={(val: string) => setValue(val)}
        onSearch={onSearch(router, 'name', { ...initialCostModelsQuery, ...query }, { value, setValue })}
      />
    ) : null;
  return (
    <ToolbarFilter
      deleteLabel={deleteChip}
      deleteLabelGroup={deleteChipGroup}
      labels={chips}
      categoryName={categoryName}
    >
      {children}
    </ToolbarFilter>
  );
};
const NameFilterConnect = connect(nameFilterMapStateToProps, undefined, nameFilterMergeProps)(NameFilterBase);
export const NameFilter = injectIntl(withRouter(NameFilterConnect));

interface SourceTypeFilterOwnProps {
  deleteChip?: any;
  deleteChipGroup?: any;
  filterType?: any;
  chips?: any;
  categoryName?: string;
  query?: any;
}

type SourceTypeFilterProps = SourceTypeFilterOwnProps & WrappedComponentProps & RouterComponentProps;

const sourceTypeFilterMapStateToProps = (state: RootState) => {
  const filterType = costModelsSelectors.currentFilterType(state);
  const query: Partial<CostModelsQuery> = costModelsSelectors.query(state);
  return { filterType, query };
};

const sourceTypeFilterMergeProps = (
  stateProps: ReturnType<typeof sourceTypeFilterMapStateToProps>,
  _dispatchProps,
  ownProps: WrappedComponentProps & RouterComponentProps
) => {
  const {
    intl = defaultIntl, // Default required for testing
    router,
  } = ownProps;
  const { filterType, query } = stateProps;

  const chips = query.source_type ? [query.source_type] : [];
  return {
    chips,
    categoryName: intl.formatMessage(messages.sourceType),
    deleteChip: onDeleteChipGroup(router, { ...initialCostModelsQuery, ...query }, 'source_type'),
    filterType,
    intl,
    query,
    router,
  };
};

const SourceTypeFilterBase: React.FC<SourceTypeFilterProps> = ({
  deleteChip,
  deleteChipGroup,
  filterType,
  intl,
  chips,
  categoryName,
  query,
  router,
}) => {
  const [currentItem, setCurrentItem] = useState<string>();

  const handleOnSelect = (_evt, selection: SelectWrapperOption) => {
    setCurrentItem(selection.value);
    router.navigate(stringifySearch({ ...initialCostModelsQuery, ...query, source_type: selection.value }));
  };

  const selectOptions: SelectWrapperOption[] = [
    {
      toString: () => intl.formatMessage(messages.aws),
      value: 'aws',
    },
    {
      toString: () => intl.formatMessage(messages.azure),
      value: 'azure',
    },
    {
      toString: () => intl.formatMessage(messages.openShift),
      value: 'ocp',
    },
  ];
  const selection = selectOptions.find(option => option.value === currentItem);

  const children =
    filterType === 'sourceType' ? (
      <SelectWrapper
        id="source-type-select"
        onSelect={handleOnSelect}
        options={selectOptions}
        placeholder={intl.formatMessage(messages.filterByPlaceholder, { value: 'source_type' })}
        selection={selection}
      />
    ) : null;
  return (
    <ToolbarFilter
      deleteLabel={deleteChip}
      deleteLabelGroup={deleteChipGroup}
      labels={chips}
      categoryName={categoryName}
    >
      {children}
    </ToolbarFilter>
  );
};

const SourceTypeFilterConnect = connect(
  sourceTypeFilterMapStateToProps,
  undefined,
  sourceTypeFilterMergeProps
)(SourceTypeFilterBase);
export const SourceTypeFilter = injectIntl(withRouter(SourceTypeFilterConnect));

const toolbarMapStateToProps = (state: RootState) => {
  const query: Partial<CostModelsQuery> = costModelsSelectors.query(state);
  return { query };
};

const toolbarMergeProps = (
  stateProps: ReturnType<typeof toolbarMapStateToProps>,
  _dispatchProps: unknown,
  ownProps: RouterComponentProps & { children: React.ReactNode }
) => {
  const { children, router } = ownProps;
  const { query } = stateProps;
  return {
    id: 'cost-models-toolbar',
    clearAllFilters: () =>
      router.navigate(
        stringifySearch({ ...initialCostModelsQuery, ...query, description: null, source_type: null, name: null })
      ),
    children,
  } as ToolbarProps;
};

export const ClearableToolbar = withRouter(connect(toolbarMapStateToProps, undefined, toolbarMergeProps)(Toolbar));
