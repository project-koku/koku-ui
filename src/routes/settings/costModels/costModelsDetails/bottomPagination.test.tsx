import { render, screen } from '@testing-library/react';
import { emptyPage, page1 } from 'api/costModels.data';
import React from 'react';
import { Provider } from 'react-redux';
import { createMemoryRouter, RouterProvider } from 'react-router-dom';
import { createStore } from 'redux';
import { FetchStatus } from 'store/common';
import type { RootState } from 'store/rootReducer';
import { rootReducer } from 'store/rootReducer';

import { CostModelsBottomPagination } from './bottomPagination';
import { initialCostModelsQuery } from './utils/query';

const renderUI = (state: Partial<RootState>) => {
  const store = createStore(rootReducer, state);
  const router = createMemoryRouter(
    [
      {
        path: '/',
        element: <CostModelsBottomPagination />,
      },
    ],
    {
      // Set for where you want to start in the routes. Remember, KISS (Keep it simple, stupid) the routes.
      initialEntries: ['/'],
      // We don't need to explicitly set this, but it's nice to have.
      initialIndex: 0,
    }
  );
  return render(
    <Provider store={store}>
      <RouterProvider
        future={{
          v7_relativeSplatPath: true,
          v7_startTransition: true,
        } as any}
        router={router}
      />
    </Provider>
  );
};

test('first page pagination', () => {
  const state = {
    costModels: {
      costModels: page1,
      status: FetchStatus.complete,
      error: null,
      query: initialCostModelsQuery,
    },
  };
  renderUI(state);
  expect(screen.queryAllByText(/1 - 10/i)).toHaveLength(1);
  expect(screen.queryAllByText(/20/i)).toHaveLength(1);
});

test('empty page pagination', () => {
  const state = {
    costModels: {
      costModels: emptyPage,
      status: FetchStatus.complete,
      error: null,
      query: initialCostModelsQuery,
    },
  };
  renderUI(state);
  expect(screen.queryAllByText(/1 - 10/i)).toHaveLength(0);
  expect(screen.queryAllByText(/0/i)).toMatchSnapshot();
});
