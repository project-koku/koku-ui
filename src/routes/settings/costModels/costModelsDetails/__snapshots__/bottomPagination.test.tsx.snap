// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`empty page pagination 1`] = `
[
  <b>
    0
     - 
    0
  </b>,
  <b>
    0
  </b>,
  <span
    aria-hidden="true"
  >
    of
     
    0
  </span>,
]
`;
